### PR TITLE
Improved packaging mark 7

### PR DESCRIPTION
**Pull request by williamjameshandley**
# Description

This addresses the concern raised [here](https://github.com/williamjameshandley/anesthetic/commit/9d7c544e1b2493c07c5b9cf1cb2d860d8e1a92d3#r102892464), by ensuring that the version number is 'unit incremented'. I believe that the updated `./bin/check_version.py` addresses all the corner cases I can see.

# Checklist:

- [X] I have performed a self-review of my own code
- [X] My code is PEP8 compliant (`flake8 anesthetic tests`)
- [X] My code contains compliant docstrings (`pydocstyle --convention=numpy anesthetic`)
- [X] New and existing unit tests pass locally with my changes (`python -m pytest`)
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have appropriately incremented the [semantic version number](https://semver.org/) in both README.rst and anesthetic/_version.py

### CODE REVIEW

**Comment by lukashergt** [comment]
Do we want to enforce strict unit increments? 

I personally don't see a problem in a skipped version number. As long as the versions are monotonically increasing it is fine with me. 

I see an opportunity in allowing non-unit increments in addressing https://github.com/williamjameshandley/anesthetic/pull/262#issuecomment-1426997940. If we have multiple PRs going in parallel and we have an idea which order we intend to merge them, then we can choose to set a higher version number for a second PR, approve it, merge a first, update the second, merge the second which still passes.

Ok, continuing my stream of consciousness. Maybe all this could work with unit increments, too, by approving a failing PR, which through an update could become passing... Could that work? But in that case as a reviewer I would want to be certain that the version number is the only thing that is failing. So to that end I would want a dedicated job to the version number, allowing me to see that lint, sphinx, pip, conda, etc. all pass and only the version job fails. I think a dedicated version job would be helpful in any case, such that we don't even need to look into the lint check to know what is going wrong.

**Comment by Ormorod** [comment]
I like having a dedicated version check too.

**Review by Ormorod — CHANGES_REQUESTED** [review]

File: bin/check_version.py [hunk 1]
```diff
@@ -24,11 +26,64 @@ def run(*args):
...
+
+for a, b in [['2.0.0b3', '2.0.0b1'],
+             ['2.0.0b3', '2.0.0a3'],
+             ['2.0.1', '2.0.0b1'],
+             ['3.0.0a1', '2.5.6'],
+             ['2.0.4', '2.0.2'], 
+             ['2.1.5', '2.0.5'],
+             ['3.5.6', '2.5.6'],
+             ]:
+    assert not unit_incremented(version.parse(a), version.parse(b))
```
[the remark this answers is not in the document]
Comment (williamjameshandley) [reply bin/check_version.py]: Now moved into tests

**Comment by williamjameshandley** [comment]
I have also used this PR to fix the bug found in #273 by @ThomasGesseyJones (and also Metha Prathaban a few weeks before)

**Comment by williamjameshandley** [comment]
>Maybe all this could work with unit increments, too, by approving a failing PR, which through an update could become passing... Could that work? But in that case as a reviewer I would want to be certain that the version number is the only thing that is failing. So to that end I would want a dedicated job to the version number, allowing me to see that lint, sphinx, pip, conda, etc. all pass and only the version job fails. I think a dedicated version job would be helpful in any case, such that we don't even need to look into the lint check to know what is going wrong.

Agree. Let's try it and see -- we can remove the requirement if it's too onerous.

**Comment by Ormorod** [comment]
I know it's inconvenient to make a new PR to fix each individual issue, but it makes it much easier to work with in the future (I have learned this the hard way!)

**Comment by Ormorod** [comment]
`tests/test_scripts.py` is now failing locally with `ModuleNotFoundError: No module named 'bin'`, which I can fix with adding a blank `.bin/__init__.py`. @williamjameshandley are you able to reproduce this?

**Comment by williamjameshandley** [comment]
It works locally for me. Can you run:
```bash
git clone git@github.com:williamjameshandley/anesthetic.git
cd anesthetic
python -m venv venv
source venv/bin/activate
pip install .'[test]'
python -m pytest -k incremented
```
and confirm whether or not this runs through?

Ideally there shouldn't be an `__init__.py`, since this would make it a module.

(for the record, this is the reason why I had the tests embedded in the script in the first place).

**Comment by Ormorod** [comment]
> (for the record, this is the reason why I had the tests embedded in the script in the first place).

I did see the irony of my requesting a change and it immediately breaking on my end 😂

> ```shell
> python -m pytest -k incremented
> ```
> 
> and confirm whether or not this runs through?
> 

it runs - it seems my issue was simply writing
```shell
pytest
```
instead of 
```shell
python -m pytest
```
What happens if you just write pytest?

**Comment by williamjameshandley** [comment]
> What happens if you just write pytest?

```bash
ModuleNotFoundError: No module named 'anesthetic'
```
Perhaps just running pytest runs it on your global install?

**Comment by Ormorod** [comment]
don't think so, as ```pytest``` isn't found if I don't activate a venv (I keep my global python completely fresh and empty)

But yes I'll assume this is a me/macOS(/maybe python3.11?) problem, please press on
